lib/fluent/plugin/elasticsearch_error_handler.rb

```diff
@@ -4,20 +4,22 @@ class Fluent::ElasticsearchErrorHandler
   include Fluent::ElasticsearchConstants
 
   attr_accessor :records, :bulk_message_count
-  class BulkIndexQueueFull < StandardError; end
-  class ElasticsearchOutOfMemory < StandardError; end
   class ElasticsearchVersionMismatch < StandardError; end
-  class UnrecognizedElasticsearchError < StandardError; end
   class ElasticsearchError < StandardError; end
+
   def initialize(plugin, records = 0, bulk_message_count = 0)
     @plugin = plugin
     @records = records
     @bulk_message_count = bulk_message_count
   end
 
-  def handle_error(response)
+  def handle_error(response, tag, records)
+    if records.length != response['items'].length
+      raise ElasticsearchError, "The number of records submitted do not match the number returned. Unable to process bulk response"
+    end
+    retry_records = []
     stats = Hash.new(0)
-    response['items'].each do |item|
+    response['items'].each_with_index do |item, index|
```

Comment:
also iterate `chunk` - pull the next item from `chunk`
You'll also have to implement the `        next unless record.is_a? Hash ` logic, and any logic in `process_message` that omits the record from the bulk index request.  That will need to be refactored into a method callable from both places.  You have to guarantee that the record you read from the `chunk` is the same one being processed in this error handling loop.

Reply:
Same here.  We've already processed the chunk.  There is no reason to do so here too.